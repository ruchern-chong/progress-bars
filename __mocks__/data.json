{
  "buttons": [
    20,
    22,
    -8,
    -32
  ],
  "bars": [
    47,
    51,
    13
  ],
  "limit": 180
}