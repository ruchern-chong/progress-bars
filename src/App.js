import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import ButtonGroup from '@material-ui/core/ButtonGroup';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Select from '@material-ui/core/Select';
import Typography from '@material-ui/core/Typography';

import ProgressBar from './components/ProgressBar';

import { fetchData, setBarValue } from './actions';

export const App = props => {
  const { bars, buttons, limit } = props;

  const [selectedBar, setSelectedBar] = useState(0);

  useEffect(
    () => {
      props.fetchData();
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const hasData = Boolean(bars && buttons && limit);

  return (
    <Container fixed>
      {hasData && (
        <Grid container justify="center" alignItems="center" spacing={2}>
          <Grid item xs={12}>
            <Typography variant="h3">Progress Bar Demo</Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">{`Limit ${limit}%`}</Typography>
          </Grid>
          {bars.map((bar, index) => (
            <Grid key={index} item xs={12}>
              {`#${index + 1}`} <ProgressBar bar={bar} limit={limit} />{' '}
              <span
                data-testid={`progress-bar-${index + 1}-value`}
              >{`${bar}%`}</span>
            </Grid>
          ))}
          <Grid item xs={12}>
            <Select
              native
              name="select"
              id="select"
              onChange={e => setSelectedBar(e.target.value)}
              data-testid="select"
            >
              {bars.map((bar, index) => (
                <option key={index} value={index}>
                  Progress Bar {index + 1}
                </option>
              ))}
            </Select>
          </Grid>
          <Grid item xs={12}>
            <ButtonGroup>
              {buttons.map((button, index) => (
                <Button
                  key={index}
                  variant="outlined"
                  onClick={() => props.setBarValue(Number(selectedBar), button)}
                  data-testid={`button`}
                >
                  {button}
                </Button>
              ))}
            </ButtonGroup>
          </Grid>
        </Grid>
      )}
    </Container>
  );
};

App.defaultProps = {
  bars: [],
  buttons: [],
  limit: 0,
  fetchData: () => {},
  setBarValue: () => {}
};

App.propTypes = {
  bars: PropTypes.arrayOf(PropTypes.number),
  buttons: PropTypes.arrayOf(PropTypes.number),
  limit: PropTypes.number,
  fetchData: PropTypes.func,
  setBarValue: PropTypes.func
};

/* istanbul ignore next */
const mapStateToProps = state => ({
  bars: state.bars,
  buttons: state.buttons,
  limit: state.limit
});

/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchData: () => dispatch(fetchData()),
  setBarValue: (selectedBar, value) => dispatch(setBarValue(selectedBar, value))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
