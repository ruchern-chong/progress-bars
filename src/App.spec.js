import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { App } from './App';

import { bars, buttons, limit } from '../__mocks__/data.json';

describe('App', () => {
  test('should render', () => {
    const { container } = render(
      <App bars={bars} buttons={buttons} limit={limit} fetchData={() => {}} />
    );

    expect(container).toMatchSnapshot();
  });

  // TODO Fix the test
  test('should fire the click event when button is clicked', () => {
    const setBarValue = jest.fn();
    const { getByTestId, getByText } = render(
      <App
        bars={bars}
        buttons={buttons}
        limit={limit}
        setBarValue={setBarValue}
      />
    );

    expect(getByTestId('progress-bar-1-value')).toHaveTextContent('47%');
    fireEvent.click(getByText('20'));
    expect(setBarValue).toBeCalledWith(0, 20);
    // expect(getByTestId('progress-bar-1-value')).toHaveTextContent('67%');
  });

  test('should change the selected progress bar', () => {
    const { getByText } = render(
      <App bars={bars} buttons={buttons} limit={limit} />
    );

    fireEvent.click(getByText('Progress Bar 1'));
    fireEvent.click(getByText('Progress Bar 2'));
    fireEvent.change(getByText('Progress Bar 2').parentElement, {
      target: {
        value: 1
      }
    });

    expect(getByText('Progress Bar 2')).toBeVisible();
  });
});
