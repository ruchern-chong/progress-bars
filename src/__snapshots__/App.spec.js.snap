// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3"
        >
          Progress Bar Demo
        </h3>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          Limit 180%
        </h6>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        #1
         
        <progress
          max="100"
          value="47"
        />
         
        <span
          data-testid="progress-bar-1-value"
        >
          47%
        </span>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        #2
         
        <progress
          max="100"
          value="51"
        />
         
        <span
          data-testid="progress-bar-2-value"
        >
          51%
        </span>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        #3
         
        <progress
          max="100"
          value="13"
        />
         
        <span
          data-testid="progress-bar-3-value"
        >
          13%
        </span>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline"
          data-testid="select"
        >
          <select
            class="MuiSelect-root MuiSelect-select MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
            id="select"
            name="select"
          >
            <option
              value="0"
            >
              Progress Bar 1
            </option>
            <option
              value="1"
            >
              Progress Bar 2
            </option>
            <option
              value="2"
            >
              Progress Bar 3
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          class="MuiButtonGroup-root"
          role="group"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedOutlined MuiButton-outlined"
            data-testid="button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              20
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedOutlined MuiButton-outlined"
            data-testid="button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              22
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedOutlined MuiButton-outlined"
            data-testid="button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              -8
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedOutlined MuiButton-outlined"
            data-testid="button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              -32
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
